drafts: Add functionality to bulk delete drafts.

sbansal1999 opened the pull request:
This PR adds functionality to bulk delete drafts quickly.

Special thanks to @juliaBichler01 for working on this one at #19943.

Fixes #19360.

- The `Select All Drafts` checkbox doesn't align with other checkboxes if a scrollbar is present on the drafts window. That can be fixed by using a function like this:
```
function adjust_select_drafts_button() {
    const $drafts_list = $(".drafts-list");
    const scroll_bar_width = $drafts_list[0].offsetWidth - $drafts_list[0].clientWidth;
    $(".select-drafts-button").css("margin-right", 25 + scroll_bar_width + "px");
}
```
But I am unsure if that's the best way to do it.

- I am not sure if I need to add any node/puppeteer tests for this one.


<hr>

**Light Theme:**
<details>
<summary>Delete All Drafts</summary>


![Delete All Drafts](https://github.com/zulip/zulip/assets/35286603/8e2dc559-a23e-4459-9d04-165132daa1df)



</details>

<details>
<summary>Delete Some Drafts</summary>

![Delete Some Messages](https://github.com/zulip/zulip/assets/35286603/e947b719-1bd8-4b91-87fa-92b558120e66)


</details>

<details> 
<summary>Phone View</summary>

![Screenshot from 2023-06-14 02-46-12](https://github.com/zulip/zulip/assets/35286603/8f49fed9-39a3-452a-beee-94743c3824de)


</details>



**Dark Theme:**
<details>
<summary>Delete All Drafts</summary>


![Delete All Drafts - Dark](https://github.com/zulip/zulip/assets/35286603/1708ec13-b9fd-4a3e-a937-0382ad2e12ed)




</details>

<details>
<summary>Delete Some Drafts</summary>
 

![Delete Some Messages - Dark (1)](https://github.com/zulip/zulip/assets/35286603/c57a8b32-f391-452d-9635-2a18a5feb76e)

 
</details>


<hr>

- The PR originally didn't make use of flexboxes due to which various values were hard-coded which affected the responsiveness of the module, so made use of flexboxes wherever possible.
- Also, made some small changes to the spacing and margins to make the whole thing more nicer.
- Replaced the checkbox used in the `Select all drafts` button to the one which is being used in all the Drafts to select them.
- I have removed the `node-tests` because I am unsure if they are required or not.

<hr>


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Thanks! Other than the rebase, can you comment on what is different on this PR vs. https://github.com/zulip/zulip/pull/19943?

alya commented:
Let's also add a bit of space between "Select all drafts" and the checkbox -- they look really squished together.

alya commented:
I think you can also post in #design for further feedback; I'd include a screen capture to show how things wrap at different screen sizes.

karlstolley commented on web/templates/draft_table_body.hbs:
```diff
@@ -6,10 +6,26 @@
+                                <div class="fa fa-trash-o fa-lg" aria-hidden="true"></div>
+                            </button>
+                        </div>
+                        <button class="button small rounded select-drafts-button" type="button">
```
I'm not feeling good about this use of a `<button>` with an embedded `<label>` wrapping the actual control--the `<input>`--which then gets an `aria-hidden` attribute. I understand that we do some custom checkbox things with the `<span>`, but those are not going to be as accessible as the native checkbox controls.

As I noted [on CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/bulk.20delete.20drafts.20.2319943/near/1585514), the button pattern is causing some unpredictable behavior for keyboard navigation as well.

Is there a reason to group these elements under a button? Especially because a properly associated `<label>` element will also function to activate the input (though, again, that might be short-circuited by the use of the inner `<span>` for displaying that control)?

karlstolley commented on web/styles/drafts.css:
```diff
@@ -91,6 +138,11 @@
                 }
             }
+
+            .checkbox {
```
Rather than the fiddly `margin` values here, I wonder if you couldn't--as a separate commit--work up the `message_top_line` div to be a flexbox? Right now, these buttons rely on `inline-block`, which means that it's actual space characters in the HTML that keep them separate from each other. With a flexbox, there'd be more robust control over this bit of microlayout--which seems more important now that there are three controls here instead of two.

sbansal1999 commented:
@karlstolley, I have updated the PR to accommodate the required changes. I have also updated the PR description to reflect the new changes in the screen recording.

- I fixed the first comment regarding having an `input` inside a `button` by using an `icon` that resembles a checkbox and made the `button` behave like a checkbox. CZO Discussion: [Link](https://chat.zulip.org/#narrow/stream/101-design/topic/bulk.20delete.20drafts.20.2319943/near/1587378).

- I had to remove some classes that resulted in identical styles, which was done in the first commit. These changes were omitted in [this](https://github.com/zulip/zulip/pull/25131/commits/5edc8fc651fa0b161c0f700acbb044af0dc45fcf) commit. CZO Discussion: [Link](https://chat.zulip.org/#narrow/stream/6-frontend/topic/drafts.20css).

- The only minor issue I discovered is that when focusing on the `Select all drafts` and the `Delete Drafts` icon using the keyboard, pressing the `Spacebar` works perfectly, but pressing `Enter/Return` doesn't behave as expected. Instead, it opens the currently selected draft. I believe this issue is related to how the `Enter/Return` key press is handled in [hotkey.js](https://github.com/zulip/zulip/blob/main/web/src/hotkey.js#LL434C3-L434C3).

- Apart from the above, I [asked](https://chat.zulip.org/#narrow/stream/101-design/topic/bulk.20delete.20drafts.20.2319943/near/1589871) for feedback on the background color in the dark theme for the two buttons, with and without focus.

- Additionally, I extracted the jQuery elements used in `update_bulk_delete_ui` to their own variables to improve readability.

karlstolley commented on web/src/tippyjs.js:
```diff
@@ -143,6 +143,35 @@ export function initialize() {
 
+    delegate("body", {
+        target: ".mark-draft-tooltip",
+        delay: [400, 20],
```
There are some constants at the top of the file, like `LONG_HOVER_DELAY`, that you should use in lieu of hard-coded values like this.

karlstolley commented on web/src/tippyjs.js:
```diff
@@ -143,6 +143,35 @@ export function initialize() {
+        delay: [400, 20],
+        appendTo: () => document.body,
+        onShow(instance) {
+            let content = "Select draft";
```
Strings like this (`"Select draft"` and so on) need to be marked for translation; have a look at some of the other `onShow()` calls in the file to see how that's done with the ` $t({defaultMessage: ''});` syntax.

karlstolley commented:
This is looking really good, @sbansal1999. A handful of things I noticed from testing:

* I think it might be better to increase the tooltip delay on the individual checkboxes. At least with my mouse behavior, about the time I'd targeted the box to click, the tooltip would flash for just an instant--that might leave users at least initially wondering what they'd missed. The other two icons' tooltips might benefit from a longer delay, too, especially the trashcan (the edit one less so, since it moves to a completely different view). See the inline comment I left about setting delay times with established constants.
* Your solution to use an icon on the Select all drafts button was smart. I wish that all of the checkboxes on the page looked like that, as I think the inverse background is a strong design choice: it would make it much easier to see on a long list of drafts which ones are selected, and which ones aren't.
* The 2px outline on the active checkboxes looks too thick to me, especially contrasted against the message outline that it visually toggles back and forth against.

sbansal1999 commented:
Thanks for reviewing my work. :heart:  I will make the necessary changes and will push them soon.

sbansal1999 commented on web/styles/app_components.css:
```diff
@@ -970,7 +970,10 @@ div.overlay {
 
         .overlay_message_controls {
-            display: inline-block;
+            display: inline-flex;
```
You are right this appears to work when we set `display: flex`. I think since the whole structure is maintained using a grid it kind of makes  `flex` and `inline-flex` behave in a similar fashion. 

I will change it to `display: flex` just to keep it simple.

sbansal1999 commented:
@karlstolley I have updated the PR and I think it is ready for review.

I have made the following changes to the PR:

- The checkboxes which were earlier an `input` tag are now an icon that looks like a checkbox -- had to create some helper functions to cut out the redundant work.
- The tippy delay for the `Delete Draft` icon has been changed to `LONG_HOVER_DELAY` 
- Added `message_control_button` class to the `Trash Bin` icon so that the tippy is removed when the user tabs out of the button.

Let me know if any more changes are required in this one.

sbansal1999 commented:
Hey @karlstolley, bumping this one just in case you might have missed it.

karlstolley commented on web/styles/drafts.css:
```diff
@@ -1,103 +1,81 @@
-                    opacity: 1;
-                }
-            }
+    .drafts-list {
```
Is this a stale selector? I'm not seeing a `.removed-drafts` inside of `.drafts-list` in the DOM or the template here.

karlstolley commented on web/templates/draft_table_body.hbs:
```diff
@@ -6,10 +6,23 @@
-                    <br />
-                    {{#tr}}Drafts older than <strong>{draft_lifetime}</strong> days are automatically removed.{{/tr}}
+                <div class="header-body">
+                    <div class="removed-drafts">
```
Let's change this class to something like `removed-drafts-message` or something. As it stands, it sounds like this class marks drafts that have been removed (which of course is impossible, but future contributors will thank you).

sbansal1999 replied:
Yes, that makes perfect sense to me.

karlstolley commented on web/styles/drafts.css:
```diff
@@ -1,103 +1,81 @@
-                color: hsl(357deg 52% 57%);
-                opacity: 0.7;
+        @media (width < $lg_min) {
+            flex-direction: column;
```
I'm not certain that this ever gets applied? And if it doesn't, there's no need to set the `row` flex-direction above.

sbansal1999 replied:
This style gets applied on smaller screens -- which then leads to the `Remove Drafts Message` and the `Deleting Drafts` group being shown in a column-wise fashion.

<details>

<summary>900px Width</summary>

![image](https://github.com/zulip/zulip/assets/35286603/43d3f14c-1ff1-4a0c-86d6-cf6aa864b51c)

</details>

<details>

<summary>1000px Width</summary>

![image](https://github.com/zulip/zulip/assets/35286603/1fdf1e3e-7c06-4ca1-8562-664b98cd74c3)

</details>

I am not sure if I am missing something here.

karlstolley replied:
Strange. I don't think I was seeing that layout at 900px that your screenshot shows? But I do see it now.

While this works, it's probably also possible just to leave this layout to display in their normal flow as blocks and only introduce flexbox at `width > $lg_min`, too.

sbansal1999 replied:
Hmm yeah, that makes sense. As I can see that there is no need for a flexbox at `width <$lg_min`.

karlstolley commented on web/styles/drafts.css:
```diff
@@ -1,103 +1,81 @@
+        .removed-drafts {
+            display: block;
+            text-align: center;
+            font-size: 1em;
```
It shouldn't be necessary to set the font-size here to 1em, which basically says the text should be the same size it would be anyway.

karlstolley commented:
This is coming along nicely, @sbansal1999. Please see my inline comments, and do an audit of all of your lines of CSS and make sure that a) they're in selectors that actually are applied, and b) that they have a meaningful effect.

karlstolley commented on web/styles/drafts.css:
```diff
@@ -1,103 +1,68 @@
+
+    .mark-draft {
+        margin-top: 5px;
+        width: 10px;
```
I don't think you need this width here; testing this on a related PR this morning, that knocks the mark-draft icon out of alignment with the rest of the grid. Try removing and see if it's okay?

karlstolley replied:
Oooh, yeah, that's not good.

If you render the grid lines around the area though, on `.messagebox-content`, you'll see that the checkbox winds up too far to the right of the lines. So some other fix that keeps those boxes in line AND prevents the shift with the layout buttons is what's called for here.

sbansal1999 replied:
I have changed the width to `15px` and then now the checkbox is inside the grid line. I also had to accordingly change the padding inside the "Select all drafts" button to make sure the checkbox inside it aligns with the checkboxes in the drafts.



| 15px | 10px |
|--------|--------|
| ![image](https://github.com/zulip/zulip/assets/35286603/4d42b265-2299-4621-b63b-e4acbbacea05) | ![image](https://github.com/zulip/zulip/assets/35286603/d3609cf3-9440-4fdd-a905-5d7866b4b18b) | 


I think since the size of the checkbox is ~13px and constraining it inside `10px` resulted in the checkbox crossing the lines.

karlstolley replied:
Nice! It'd be good to record the reasoning behind the 15px width in a short comment or two in your CSS itself, so future contributors understand what's happening there (or how to translate things if we change those icons into a gridded presentation, rather than flexbox)

sbansal1999 commented:
@karlstolley Thanks for reviewing my work. :heart_decoration: 

I have updated the PR and I think it's ready for a review. 

Changes made:

- Removed the `.removed-drafts` selector which wasn't being applied.
- Renamed a class name to add some more context in c639a09f0dd182245173f59e23d726d3bfb161f9.
- Adjusted the layout of `.header-body` to just display as a block instead of a flexbox -- which resulted in the layout being simpler -- only for smaller screens.
- Earlier the checkbox icon was moving out of the grid lines; modified the width of the icon to fix that. Also added a comment for that.

karlstolley commented:
Thanks for those fixes, @sbansal1999! This is looking ready for integration review and CZO burn-in, so I'm going to label it as such.

sbansal1999 commented:
Thanks for all those reviews @karlstolley. :heart:

sbansal1999 commented:
**Updates:**

- Fixed the merged conflict.

Looks like zulipbot hasn't removed the `has conflicts` label; could be some kind of an API failure,

sbansal1999 commented:
Rebased the PR to fix merge conflicts.

timabbott commented on web/src/drafts.js:
```diff
@@ -548,8 +548,8 @@ const keyboard_handling_context = {
-    row_item_selector: "draft-row",
-    box_item_selector: "draft-info-box",
+    row_item_selector: "overlay-message-row",
+    box_item_selector: "overlay-message-info-box",
```
So I worry that these changes may result in problems where the logic 

@amanagr does `messages_overlay_ui.ts` need a small refactor to ensure that it's always using these within the scope of say `drafts-list`? A lot of the selectors, like this function, do not have an obvious mechanism through which they wouldn't accidentally match the similar elements in the scheduled messages overlay.

```
export function row_with_focus(context: Context): JQuery {                                              
    const focused_item = $(`.${CSS.escape(context.box_item_selector)}:focus`)[0];                       
    return $(focused_item).parent(`.${CSS.escape(context.row_item_selector)}`);                         
}                                                                                                       
```

timabbott commented on web/src/drafts.js:
```diff
@@ -617,20 +617,20 @@ export function launch() {
 
     function setup_event_handlers() {
-        $(".restore-draft").on("click", function (e) {
+        $(".restore-overlay-message").on("click", function (e) {
```
This change is potentially wrong: `restore-overlay-message` appears in the scheduled messages overlay too. The `$("#drafts_table ...")` approach used elsewhere seems like an appropriate fix.

Please carefully audit this whole commit for this class of issue.

amanagr replied:
@sbansal1999  I think in this commit, instead of replacing `draft` with `overlay-message`, we can replace them with `.draft-row .overlay-message`.  This would ensure everything works correctly.

> This resulted in the same styles being applied twice.

@sbansal1999  Can you describe what styles you are talking about?

sbansal1999 replied:
I feel what Tim has suggested feels like a more intuitive way to me as it would clearly tell that the element belongs to the drafts table. I would love to hear your thoughts on this one @amanagr.

> Can you describe what styles you are talking about?

I am talking about the styles that were removed in the next commit: b653d1e46616b657e2f45f00f00279feeb9ae838

timabbott replied:
I think either option is workable here.

timabbott commented on web/templates/draft_table_body.hbs:
```diff
@@ -6,10 +6,23 @@
+                        {{#tr}}Drafts older than <strong>{draft_lifetime}</strong> days are automatically removed.{{/tr}}
+                    </div>
+                    <div class="delete-drafts-group">
+                        <div class="delete-selected-drafts-button-container message_control_button">
```
What's the thinking behind using the `message_control_button` class here? I worry that the reuse of this class in just this one place is more likely to result in bugs where we accidentally break this when adjusting message feed design.

sbansal1999 replied:
I thought of using this class to make sure the tippy on the button is removed after the keyboard focus is gone to some other button. But as you mentioned using this class here can lead to bugs. 

I will do the following change in [tippyjs.js](https://github.com/zulip/zulip/blob/main/web/src/tippyjs.js#L146) and remove the above mentioned class:
```diff
-    $("body").on("blur", ".message_control_button", (e) => {
-        // Remove tooltip when user is trying to tab through all the icons.
-        // If user tabs slowly, tooltips are displayed otherwise they are
-        // destroyed before they can be displayed.
-        e.currentTarget?._tippy?.destroy();
-    });
+    $("body").on(
+        "blur",
+        ".message_control_button, .delete-selected-drafts-button-container",
+        (e) => {
+            // Remove tooltip when user is trying to tab through all the icons.
+            // If user tabs slowly, tooltips are displayed otherwise they are
+            // destroyed before they can be displayed.
+            e.currentTarget?._tippy?.destroy();
+        },
+    );

```

timabbott replied:
Yeah that sounds great.

timabbott commented on web/styles/drafts.css:
```diff
@@ -1,9 +1,73 @@
         }
     }
+
+    .mark-draft {
```
What do you think about renaming this to `draft-selection-checkbox`? I feel like "mark-draft" really sounds like it's something to "mark something as a draft".

sbansal1999 replied:
Yeah, I agree with you on this one.

timabbott commented on web/styles/app_components.css:
```diff
@@ -973,7 +973,10 @@ div.overlay {
+            display: flex;
+            align-items: center;
+            justify-content: center;
+            gap: 10px;
```
Can you confirm that this doesn't make the "scheduled messages" UI that shares these classes look busted?

sbansal1999 replied:
Yes, they don't break the "scheduled messages" UI. 

The change makes the controls use a flexbox to make it easy to do modifications such as adding an icon; which was done in this PR. 

<details>

<summary> Screenshot </summary>

![image](https://github.com/zulip/zulip/assets/35286603/1074b7da-dd8d-4f28-a85f-38881f963df2)


</details>

timabbott replied:
Great!

sbansal1999 commented on web/src/drafts.js:
```diff
@@ -657,6 +687,38 @@ export function launch() {
+        $delete_selected_drafts_button.prop("disabled", false);
+        if ($unchecked_checkboxes.length === 0) {
+            toggle_checkbox_icon_state($select_state_indicator, true);
+            $select_drafts_button.attr("aria-checked", "true");
```
Yes, you are right. This can be done by modifying the `aria-checked` property of the `parent` of the checkbox that is passed to `toggle_checkbox_icon_state`.

sbansal1999 commented:
**Updates:**
- Renamed some selectors to make sure that they don't overlap with the scheduled overlay classes as highlighted in [this](https://github.com/zulip/zulip/pull/25610#discussion_r1274411499) comment.
- Removed the one-off use of the `message_control_button` class and instead used the already present class name to handle tippys during keyboard navigation.
- Renamed the `mark-draft` class to `draft-selection-checkbox` as it made more sense to what the button actually does.
- Moved the work of changing the `aria-checked` attribute for checkbox icons inside buttons to the `toggle_checkbox_icon_state` function to increase the readability and it just makes more sense for it to happen inside the function instead of somewhere else.

timabbott commented:
Merged, huge thanks for doing this @juliaBichler01 and @sbansal1999!
